### PR TITLE
Cluster refactor organized into commits

### PR DESCRIPTION
**Pull request by jhershberg**
WIP FOR EARLY REVIEW: Prepare for multiple clustering solutions. 

This PR is a first step towards implementing "flotilla data shards", that is, redis cluster nodes that do their cluster business according to flotilla clustering instead of the existing, legacy, "OSS clustering." 

In order to do achieve this it was necessary to separate (a) the generic, shared cluster elements such as the fact that there are nodes, primaries, shards, etc., and (b) the specifics of the current OSS clustering implementation such as details stemming from the current cluster's gossip protocol. 

In general, a pragmatic approach was taken leading to some mild oddities in places where we avoided making larger changes. If we want to clean these up in the future, we can and probably should.

The cluster state consists of two primary data structures, clusterState and clusterNode. For the purposes of this refactor these two structures were reworked as follows:
 * clusterState now holds just a few fields and a void* "impl" that each implementation (legacy, and in the future - flotilla) uses for its internal data.
 * clusterNode is now a purely internal construct that is exposed via an opaque "handle" with "accessor" functions such as "clusterNodeIsMaster(clusterNodeHandle node)"

The code has been reorganized into the following files:
- cluster.h - The interface between the rest of Redis and the clustering module
- cluster.c - common clustering functions that will be shared between all cluster implementations + conditional #inclusion of the specific, not-common clustering implementation.
- cluster_legacy.h - Structs and macros used by the legacy clustering implementation
- cluster_legacy.c - implementations of the not-common clustering code
- cluster_flotilla.c/h - TODO: These will have the flotilla clustering implementations

One last disclaimer: There is some 'history' in the commits that comprise this PR that contain some earlier design ideas and various minor fixes.

### CODE REVIEW

File: src/cluster.h [hunk 1]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
+ * Functions requiring per-clustering mechanism implementation: "Accessor" functions
+ */
+char* clusterNodeLastKnownIp(clusterNode *node);
+int clusterNodePort(clusterNode *node);
+clusterNode* clusterNodeGetSlaveof(clusterNode *node);
+int clusterNodeConfirmedReachable(clusterNode* node);
+int clusterNodeIsMaster(clusterNode* node);
+int clusterNodeIsSlave(clusterNode* node);
+int clusterNodeIsFailing(clusterNode* node);
+int clusterNodeTimedOut(clusterNode* node);
```
Comment (madolson) [inline src/cluster.h]: Might just want a "clusterNodeHealth", since many of these states are overlapping.

File: src/cluster.h [hunk 2]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
+int clusterNodeIsMyself(clusterNode* node);
```
Comment (madolson) [inline src/cluster.h]: If this is a handle, this would make more sense to not be per-clustering implementation, since you should be able to still do myself == node. It would be odd to have the node "change" silently underneath you. Ditto about getMyClusterNode.

File: src/cluster.h [hunk 3]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
+int clusterNodeIsNoFailover(clusterNode* node);
```
Comment (madolson) [inline src/cluster.h]: Not sure why this needs to be exposed.

File: src/cluster.h [hunk 4]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
+clusterNode* getMyClusterNode(void);
+clusterNode* getNodeBySlot(int slot);
+clusterNode* getMigratingSlotDest(int slot);
+clusterNode* getImportingSlotSource(int slot);
+int isClusterHealthy(void);
+uint16_t getClusterNodeRedirectPort(clusterNode* node, int use_pport);
+const char *getPreferredEndpoint(clusterNode *n);
+clusterNode *clusterLookupNode(const char *name, int length);
+int isClusterManualFailoverInProgress(void);
```
Comment (madolson) [inline src/cluster.h:116]: This feels wrong in the API. I would expect Redis to be driving this, because it knows the detail of the state. I would expect both clustering systems to "initiate" the failover, but they shouldn't be the one monitoring it.

File: src/cluster.h [hunk 5]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
+void clusterUpdateMyselfFlags(void);
 unsigned long getClusterConnectionsCount(void);
+void clusterUpdateMyselfHostname(void);
+void clusterUpdateMyselfAnnouncedPorts(void);
+void clusterUpdateMyselfIp(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: cluster commands
+ */
+void clusterCommand(client *c);
```
Comment (madolson) [inline src/cluster.h]: I would expect we would split this into the application APIs (readonly, readwrite, slots, shards) and then the admin APIs.

File: src/cluster.h [hunk 6]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
+int clusterNodeIsFailing(clusterNode* node);
+int clusterNodeTimedOut(clusterNode* node);
+int clusterNodeIsMyself(clusterNode* node);
+int clusterNodeIsNoFailover(clusterNode* node);
+clusterNode* getMyClusterNode(void);
+clusterNode* getNodeBySlot(int slot);
+clusterNode* getMigratingSlotDest(int slot);
+clusterNode* getImportingSlotSource(int slot);
+int isClusterHealthy(void);
+uint16_t getClusterNodeRedirectPort(clusterNode* node, int use_pport);
```
Comment (madolson) [inline src/cluster.h]: Redirect port isn't any different from normal port. If you look upstream you'll see we've refactored this mostly to be getClusterNodePlaintextPort and getClusterNodeTLSPort. That is the interface we want to expose to flotilla.

File: src/cluster.h [hunk 7]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
 void clusterPropagatePublish(robj *channel, robj *message, int sharded);
+void slotToChannelAdd(sds channel);
+void slotToChannelDel(sds channel);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: Debug and info actions
+ */
+void freeThisNodesLink(clusterNode *node);
+void freeNodeInboundLink(clusterNode *node);
+sds clusterGenNodesDescription(client *c, int filter, int use_pport);
```
Comment (madolson) [inline src/cluster.h:94]: Does flotilla even want this? Do we want to deprecate the nodes response for it? (I think so)

File: src/cluster.h [hunk 8]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
 int clusterSendModuleMessageToTarget(const char *target, uint64_t module_id, uint8_t type, const char *payload, uint32_t len);
 void clusterPropagatePublish(robj *channel, robj *message, int sharded);
+void slotToChannelAdd(sds channel);
+void slotToChannelDel(sds channel);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: Debug and info actions
+ */
+void freeThisNodesLink(clusterNode *node);
+void freeNodeInboundLink(clusterNode *node);
```
Comment (madolson) [inline src/cluster.h]: Not sure this needs to exist in flotilla.

File: src/cluster.h [hunk 9]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
+ */
+void clusterCommand(client *c);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: Pub/Sub and module messages
+ */
 int clusterSendModuleMessageToTarget(const char *target, uint64_t module_id, uint8_t type, const char *payload, uint32_t len);
 void clusterPropagatePublish(robj *channel, robj *message, int sharded);
+void slotToChannelAdd(sds channel);
+void slotToChannelDel(sds channel);
```
Comment (madolson) [inline src/cluster.h:92]: Hmm, let's discuss this a bit. I think we made a mistake with how sharded pubsub was designed, and we might want to fix that for flotilla. I think sharded pubsub should only go to primary and be replicated OR not be propagated. Then flotilla doesn't need to worry about this.

File: src/cluster.h [hunk 10]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
+ * Cluster public structs
+ *----------------------------------------------------------------------------*/
+
+typedef struct clusterNode {
+    char name[CLUSTER_NAMELEN]; /* Node name, hex string, sha1-size */
+    char shard_id[CLUSTER_NAMELEN]; /* shard id, hex string, sha1-size */
+    void* data;
+} clusterNode;
+
 typedef struct clusterState {
```
Comment (madolson) [inline src/cluster.h:61]: I want to think through the benefit of keeping the slot->node and shard->node mapping here, the "Redis topology state". It seems like it might allow a lot more of shared code between the implementations, but at the same time it might make the interface much harder since now the clustering implementations need to have a complex update operation.

**Review by madolson — COMMENTED** [review]
I thought we were going to do this more like a first party module like how TLS was implemented?

File: src/cluster.h (hunk 4 above)
Comment (madolson) [reply src/cluster.h:116]: Thought about this more, and yeah I still think there should be a way to call "start write pause" and "stop write pause" and then get rid of this special manual failover state for cluster. I could send a PR for this now I suppose. Either that, or we should expose an API like, "Stepdown as soon as you can", and the master will wait for a replica to sync and step down.

File: src/cluster.h [hunk 11]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
+clusterNode* getNodeBySlot(int slot);
+clusterNode* getMigratingSlotDest(int slot);
+clusterNode* getImportingSlotSource(int slot);
+int isClusterHealthy(void);
+uint16_t getClusterNodeRedirectPort(clusterNode* node, int use_pport);
+const char *getPreferredEndpoint(clusterNode *n);
+clusterNode *clusterLookupNode(const char *name, int length);
+int isClusterManualFailoverInProgress(void);
+int getNumSlaves(clusterNode *node);
+clusterNode* getSlave(clusterNode *node, int slave_idx);
```
Comment (madolson) [inline src/cluster.h]: I kind of dislike that we don't just have an API get to all the nodes, it seems okay to return a list and have the caller free it.

File: src/cluster.h [hunk 12]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
+ */
+void freeThisNodesLink(clusterNode *node);
+void freeNodeInboundLink(clusterNode *node);
+sds clusterGenNodesDescription(client *c, int filter, int use_pport);
+sds genClusterInfoString(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: "Accessor" functions
+ */
+char* clusterNodeLastKnownIp(clusterNode *node);
```
Comment (madolson) [inline src/cluster.h]: From an API mirroring perspective, all of this is "last known" from the classic perspective.

File: src/cluster.h [hunk 13]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
+ * Functions requiring per-clustering mechanism implementation: Debug and info actions
+ */
+void freeThisNodesLink(clusterNode *node);
+void freeNodeInboundLink(clusterNode *node);
+sds clusterGenNodesDescription(client *c, int filter, int use_pport);
+sds genClusterInfoString(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: "Accessor" functions
+ */
```
Comment (madolson) [inline src/cluster.h]: From a structure perspective, I would prefer to group all of the "per-clustering" sections together then have sub grouping.

File: src/cluster.h [hunk 14]
```diff
@@ -173,257 +51,102 @@ typedef struct {
...
-#define CLUSTERMSG_FLAG0_FORCEACK (1<<1) /* Give ACK to AUTH_REQUEST even if
-                                            master is up. */
-#define CLUSTERMSG_FLAG0_EXT_DATA (1<<2) /* Message contains extension data */
-
 /* ---------------------- API exported outside cluster.c -------------------- */
+/*
+ * Functions requiring per-clustering mechanism implementation: Lifecycle events
+ */
 void clusterInit(void);
 void clusterInitListeners(void);
```
Comment (madolson) [inline src/cluster.h]:
```suggestion
void clusterInitLast(void);
```

File: src/cluster_legacy.c [hunk 15]
```diff
@@ -0,0 +1,6318 @@
...
+        return C_OK;
+
+    /* If this node is a slave, don't perform the check at all as we
+     * completely depend on the replication stream. */
+    if (nodeIsSlave(myself)) return C_OK;
+
+    /* Make sure we only have keys in DB0. */
+    for (j = 1; j < server.dbnum; j++) {
+        if (dictSize(server.db[j].dict)) return C_ERR;
+    }
```
Comment (madolson) [inline src/cluster_legacy.c:4847]: This seems like generic code, so we probably want to crack this function into two parts.

File: src/cluster.c [hunk 16]
```diff
@@ -1,1491 +1,517 @@
...
-        dictDelete(server.cluster->nodes,oldname);
-        sdsfree(oldname);
-        getRandomHexChars(myself->name, CLUSTER_NAMELEN);
-        getRandomHexChars(myself->shard_id, CLUSTER_NAMELEN);
-        clusterAddNode(myself);
-        serverLog(LL_NOTICE,"Node hard reset, now I'm %.40s", myself->name);
+    uint64_t cmd_flags = getCommandFlags(c);
+    /* Cluster is globally down but we got keys? We only serve the request
+     * if it is a read command and when allow_reads_when_down is enabled. */
+    if (!isClusterHealthy()) {
```
Comment (madolson) [inline src/cluster.c:324]: I would posit that I think there is utility in this having 3 states:
1. I can serve all traffic. (AFAIK everything is healthy in the cluster so I can serve slots based on my known state)
2. I can serve reads. (I'm not sure about the cluster state, but I know I still own my slots)
3. I should serve no traffic. (I have no idea what is going on, maybe the cluster is completing migrated and I should serve nothing)

I think for example "cluster_allow_reads_when_down" is really a statement about 2, but flotilla might have a different perspective about when a node should be able to serve reads.

Comment (jhershberg) [reply src/cluster.c:324]: This seems reasonable. I mainly worked on a 1-1 translation/reorganization of the concepts into an interface that would allow us two separate (legacy and flotilla) imlementations, not on improvements. Of couse I am open to improvements but would prefer, not die on the hill of, keeping this PR focused as a reorganization, rather than making logical changes to the code.

File: src/cluster.h (hunk 1 above)
Comment (jhershberg) [reply src/cluster.h]: What, and like have it return flags or something? Not sure I'm following what you mean here.

File: src/cluster.h (hunk 2 above)
Comment (jhershberg) [reply src/cluster.h]: Actually, this function is used both in the "shared" cluster code (cluster.c) and in module.c (grrrr). Saying "myself == node" requires holding a static myself. Since IsMyself needs to be determined both within the clustering code and within redis (module.c), it makes more sense to me that the clustering hold the static myself and then answer questions about it.

File: src/cluster.h (hunk 3 above)
Comment (jhershberg) [reply src/cluster.h]: module.c :-(

File: src/cluster.h (hunk 4 above)
Comment (jhershberg) [reply src/cluster.h:116]: It does feel wrong but it's required in replication.c

File: src/cluster.h (hunk 5 above)
Comment (jhershberg) [reply src/cluster.h]: Yeah, I also would like to do that. Of the various cluster sub-commands...
* some make no sense in flotilla
* I generalized the cluster slots implementation and moved it into cluster.c
* There are a few more that have generic implementations that can be shared
* cluster shards is a challenge. It's implemented in a manner that makes it very hard to abstract and make shared.

Here is a cut-and-paste from some notes I took ("return error" means it is N/A in flotilla:

"ADDSLOTS <slot> [<slot> ...]", - RETURN ERROR
"ADDSLOTSRANGE <start slot> <end slot> [<start slot> <end slot> ...]", RETURN ERROR
"BUMPEPOCH", RETURN ERROR
"COUNT-FAILURE-REPORTS <node-id>", RETURN ERROR
"COUNTKEYSINSLOT <slot>", SHARED IMPLEMENT
"DELSLOTS <slot> [<slot> ...]", RETURN ERROR
"DELSLOTSRANGE <start slot> <end slot> [<start slot> <end slot> ...]", RETURN ERROR
"FAILOVER [FORCE|TAKEOVER]", ? SHORT TERM NO FORCE, LONG TERM - YES
"FORGET <node-id>", RETURN ERROR
"GETKEYSINSLOT <slot> <count>", SHARED IMPL
"FLUSHSLOTS", RETURN ERROR
"INFO", IMPLEMENT
"KEYSLOT <key>", SHARED CODE
"MEET <ip> <port> [<bus-port>]", RETURN ERROR
"MYID", IMPLEMENT
"MYSHARDID", IMPLEMENT
"NODES", IMPLEMENT BUT MOST OF THE OUTPUT IS N/A
"    Return cluster configuration seen by node. Output format:",
"    <id> <ip:port@bus-port[,hostname]> <flags> <master> <pings> <pongs> <epoch> <link> <slot> ...",
"REPLICATE <node-id>", RETURN ERROR
"RESET [HARD|SOFT]", RETURN ERROR
"SET-CONFIG-EPOCH <epoch>", RETURN ERROR
"SETSLOT <slot> (IMPORTING <node-id>|MIGRATING <node-id>|STABLE|NODE <node-id>)", RETURN ERROR
"REPLICAS <node-id>", IMPLEMENT
"SAVECONFIG", MAYBE EVENTUALLY
"SLOTS", IMPLEMENT
"SHARDS", IMPLEMENT
"LINKS", RETURN ERROR

In addition we will probably want to add commands.

In short, I totally agree with this comment and I am happy to work on doing this but the PR up to this point represents my attempt to get something ready that we can experiment with. I can do this now, in this PR, or later. What's your take?

File: src/cluster.h (hunk 6 above)
Comment (jhershberg) [reply src/cluster.h]: So this does not quite allign with what you have upstream as it uses the use_pport param to decide which port to return. That is why I named the function after its usage to try and differentiate the name from just "give me this field." Open to new names but it's neither PlaintextPort or TLSPort.

File: src/cluster.h (hunk 7 above)
Comment (jhershberg) [reply src/cluster.h:94]: No but it's in use in debug.c (also in clusterCommand()). There are a bunch things here that are only used in debug.c. One option, if we want to further hide stuff, is that the cluster could expose a clusterDebugCommand. I like that.

File: src/cluster.h (hunk 8 above)
Comment (jhershberg) [reply src/cluster.h]: This definitely does not need to exist but it is required for debug.c. See my comment about clusterDebugCommand

File: src/cluster.h (hunk 9 above)
Comment (jhershberg) [reply src/cluster.h:92]: That would be great. Since we do not have a cluster bus in Flotilla I was just procrastinating when I left this in.

File: src/cluster.h (hunk 10 above)
Comment (jhershberg) [reply src/cluster.h:61]: Yes. Totally. Also undecided about this.

File: src/cluster.h (hunk 11 above)
Comment (jhershberg) [reply src/cluster.h]: ...drum roll...

File: src/cluster.h (hunk 5 above)
Comment (jhershberg) [reply src/cluster.h]: I would like to work on this now I think. What do you mean by readonly and readwrite?

File: src/cluster.h (hunk 12 above)
Comment (jhershberg) [reply src/cluster.h]: fixed

File: src/cluster_legacy.c (hunk 15 above)
Comment (jhershberg) [reply src/cluster_legacy.c:4847]: Do you mean these lines or the whole function...The function could be generalized as well...only clusterSaveConfigOrDie would need to be specialized for the two implementations. Seems like a good idea to me. Thoughts?

File: src/cluster.h (hunk 14 above)
Comment (jhershberg) [reply src/cluster.h]: fixed

File: src/cluster.h (hunk 13 above)
Comment (jhershberg) [reply src/cluster.h]: sure :-) Fixed

File: src/cluster.h (hunk 1 above)
Comment (madolson) [reply src/cluster.h]: Either return an enum for the allowed states (my preference) or we could return some flags. Redis clients don't care about all these substates, they basically only care about "Can I talk to this node", which is what I expect the high level abstraction to vend.

File: src/module.c [hunk 17]
```diff
@@ -8906,35 +8906,34 @@ int RM_GetClusterNodeInfo(RedisModuleCtx *ctx, const char *id, char *ip, char *m
...
-        if (node->flags & CLUSTER_NODE_SLAVE) *flags |= REDISMODULE_NODE_SLAVE;
-        if (node->flags & CLUSTER_NODE_PFAIL) *flags |= REDISMODULE_NODE_PFAIL;
-        if (node->flags & CLUSTER_NODE_FAIL) *flags |= REDISMODULE_NODE_FAIL;
-        if (node->flags & CLUSTER_NODE_NOFAILOVER) *flags |= REDISMODULE_NODE_NOFAILOVER;
+        if (clusterNodeIsMyself(node)) *flags |= REDISMODULE_NODE_MYSELF;
+        if (clusterNodeIsMaster(node)) *flags |= REDISMODULE_NODE_MASTER;
+        if (clusterNodeIsSlave(node)) *flags |= REDISMODULE_NODE_SLAVE;
+        if (clusterNodeTimedOut(node)) *flags |= REDISMODULE_NODE_PFAIL;
+        if (clusterNodeIsFailing(node)) *flags |= REDISMODULE_NODE_FAIL;
+        if (clusterNodeIsNoFailover(node)) *flags |= REDISMODULE_NODE_NOFAILOVER;
```
Comment (madolson) [inline src/module.c:8936]: Actually reading this now, yeah I think the per-cluster implementation should generate these flags.

EDIT: So I get Myself, master, slave. I really don't think modules should have introspection into pfail though. I'm not sure about no-failover though and fail though. Fail makes some amount of sense, it corresponds to "unhealthy" basically. There was some discussion that "dead" should really just be not shown in the cluster though.

File: src/cluster.c [hunk 18]
```diff
@@ -1496,5982 +522,848 @@ int verifyClusterNodeId(const char *name, int length) {
...
-    } else {
-        serverPanic("getNodeByQuery() unknown error.");
-    }
-}
-
-/* This function is called by the function processing clients incrementally
- * to detect timeouts, in order to handle the following case:
- *
- * 1) A client blocks with BLPOP or similar blocking operation.
- * 2) The master migrates the hash slo
```
Comment (madolson) [inline src/cluster.c:1296]: We could drop clusterNodeIsMyself() if we fix the module.c and then also change getReplOffset() to also take the current replicationGetSlaveOffset().

File: src/cluster.h (hunk 4 above)
Comment (madolson) [reply src/cluster.h:116]: Ok, to make my position clear, I would like us to build a coherent API, even if that requires some modification to how the code is currently is implemented. If you want to split between refactoring and rewriting, that's OK too, but let's finish up both before we do too much integrating with flotilla.

File: src/cluster.h (hunk 5 above)
Comment (madolson) [reply src/cluster.h]: > What do you mean by readonly and readwrite?

My brain was just thinking about the client APIs for the cluster spec, of which `READWRITE` and `READONLY` are there but have nothing to do with the cluster command. 

> "SAVECONFIG", MAYBE EVENTUALLY

I don't know why this would be shared, flotilla shouldn't care?

> FAILOVER

Maybe we can simplify this and just return OK for the arguments we silently ignore. For example AWS used the force argument forever because we didn't really think it through. 

> NODES

We should make a decision if we want to maintain this or not. It's supposed to be an admin command, I'm fine not having it be supported in flotilla and keep pushing clients to adopt slots or shards.

Comment (madolson) [reply src/cluster.h]: > I can do this now, in this PR, or later. What's your take?

Made in another comment I see two parts, refactoring and interface implementation, I think they should be separate but not specifically concerned which one is first.

File: src/cluster.h (hunk 6 above)
Comment (madolson) [reply src/cluster.h]: Let me try to add more color. The messaging the clusterbus sends around has a pport, because that is how it used to work so that is backwards compatible. The cluster node object has no pport, it just has TLS and plaintext. The decision we made is that we want the higher abstraction, the node, to just expose TLS and plaintext. See https://github.com/redis/redis/blob/unstable/src/cluster.c#L114.

File: src/cluster.h (hunk 7 above)
Comment (madolson) [reply src/cluster.h:94]: > One option, if we want to further hide stuff, is that the cluster could expose a clusterDebugCommand. I like that

I also like this.

File: src/cluster.h (hunk 8 above)
Comment (madolson) [reply src/cluster.h]: Ack, agree with earlier suggestion. This is all just for test writing anyways.

File: src/cluster.h (hunk 9 above)
Comment (madolson) [reply src/cluster.h:92]: I guess let me make an issue or something about this in the main thread. We can just leave this for now then.

File: src/cluster.h (hunk 10 above)
Comment (madolson) [reply src/cluster.h:61]: I spent like an hour prototyping it, and my conclusion is probably this is fine. Basically:
**Pros:***
- Cluster shards, slots and some other misc commands become much cleaner, since we have all the state to show them.
- The API contract is smaller by a couple of APIs. We can compact a lot of stuff together in the updates, not sure if that is a huge win.

**Cons:**
- We basically need inverse APIs for the node update operations, which are arguably more annoying to orchestrate since you have to understand when to call them. 
- Each implementation still needs it's own map of data, so there is quite a bit of duplication. The data doesn't need to be same I suppose. Duplication of data is worse than duplication of code IMO, lower chance of data drift.

File: src/cluster.h (hunk 3 above)
Comment (madolson) [reply src/cluster.h]: Hmm, that seems like a bad decision, maybe we can deprecate it.

File: src/cluster.h (hunk 10 above)
Comment (madolson) [reply src/cluster.h:61]: We might be able to just add some APIs anyway to solve the shards/slots issues, so maybe that is preferable anyways. getShards() and getNodesFrom Shard().

Comment (jhershberg) [reply src/cluster.h:61]: "CLUSTER SHARDS" is already in cluster.c and works with the current exposed API. "CLUSTER SLOTS" on the other hand is very enmeshed in the internal "legacy" implementation, partially for optimization. In the mean time I've left it where it is assuming that we will either create a single more generic implementation later or flotilla will implement it for itself from scratch.

File: src/cluster.h (hunk 4 above)
Comment (jhershberg) [reply src/cluster.h:116]: Sorry for the crossed wires but my response was to your initial comment here, not the second one. I think I did not refresh my browser or something when I wrote it. Anyway, your proposal does sound like a good idea though I would prefer to keep this PR focused on refactoring.

File: src/cluster.c (hunk 18 above)
Comment (jhershberg) [reply src/cluster.c:1296]: clusterNodeIsMyself is currently in use in module.c as well. I think it is worth leaving in place the bits required to maintain naive support of the current module.c stuff and then deprecating the whole concept and at that point removing the parts of the cluster API that are not longer required.

File: src/cluster.h (hunk 10 above)
Comment (jhershberg) [reply src/cluster.h:61]: Also, the main parts of the cluster implementation, the parts that serve data, are currently generic implementations in cluster.c. Would it be nice to have more shared code? Definitely, agree with your cost/benefit analysis that it's not worth it in this case.

File: src/cluster.h (hunk 4 above)
Comment (madolson) [reply src/cluster.h:116]: Cool, we had a brief talk with the core team and agreed that early in Redis 8 was the time for refactoring, so we should be able to get all this stuff done.

File: src/cluster.c (hunk 18 above)
Comment (madolson) [reply src/cluster.c:1296]: Ack

File: src/cluster_legacy.c [hunk 19]
```diff
@@ -6234,3 +6234,16 @@ int handleDebugClusterCommand(client *c) {
 
     return 1;
 }
+
+int clusterAllowFailoverCmd(client *c) {
```
Comment (madolson) [inline src/cluster_legacy.c:6230]: Why is this per implementation? I assume we would always direct clients to the cluster failover command?

Comment (jhershberg) [reply src/cluster_legacy.c:6230]: Correct. It returns true in the flotilla implementation. It's just a simple way to keep the call site generic.
